model/atmosphere/dycore/tests/dycore_tests/test_solve_nonhydro.py

```diff
@@ -1755,3 +1740,421 @@ def test_apply_divergence_damping_and_update_vn(
+        current_exner=current_exner,
+        current_rho=current_rho,
+        current_theta_v=current_theta_v,
+        current_w=current_w,
+        inv_ddqz_z_full=metrics_savepoint.inv_ddqz_z_full(),
+        exner_w_implicit_weight_parameter=metrics_savepoint.vwind_impl_wgt(),
+        theta_v_at_cells_on_half_levels=theta_v_at_cells_on_half_levels,
+        perturbed_exner_at_cells_on_model_levels=perturbed_exner_at_cells_on_model_levels,
+        exner_tendency_due_to_slow_physics=exner_tendency_due_to_slow_physics,
+        rho_iau_increment=rho_iau_increment,
+        exner_iau_increment=exner_iau_increment,
+        ddqz_z_half=metrics_savepoint.ddqz_z_half(),
+        rayleigh_damping_factor=rayleigh_damping_factor,
+        reference_exner_at_cells_on_model_levels=metrics_savepoint.exner_ref_mc(),
+        advection_explicit_weight_parameter=advection_explicit_weight_parameter,
+        advection_implicit_weight_parameter=advection_implicit_weight_parameter,
+        lprep_adv=savepoint_nonhydro_init.get_metadata("prep_adv").get("prep_adv"),
+        r_nsubsteps=r_nsubsteps,
+        ndyn_substeps_var=float(config.ndyn_substeps_var),
+        iau_wgt_dyn=iau_wgt_dyn,
+        dtime=savepoint_nonhydro_init.get_metadata("dtime").get("dtime"),
+        is_iau_active=is_iau_active,
+        rayleigh_type=config.rayleigh_type,
+        at_first_substep=at_first_substep,
+        at_last_substep=at_last_substep,
+        index_of_damping_layer=grid_savepoint.nrdmax(),
+        jk_start=jk_start,
+        kstart_moist=kstart_moist,
+        horizontal_start=start_cell_nudging,
+        horizontal_end=end_cell_local,
+        vertical_start=0,
+        vertical_end=icon_grid.num_levels + 1,
+        offset_provider=offset_provider,
+    )
+
+    assert helpers.dallclose(
+        vertical_mass_flux_at_cells_on_half_levels.asnumpy(),
+        z_contr_w_fl_l_ref.asnumpy(),
+        atol=1e-12,
+    )
+    assert helpers.dallclose(
+        tridiagonal_beta_coeff_at_cells_on_model_levels.asnumpy(), z_beta_ref.asnumpy()
+    )
+    assert helpers.dallclose(
+        tridiagonal_alpha_coeff_at_cells_on_half_levels.asnumpy(), z_alpha_ref.asnumpy()
+    )
+    assert helpers.dallclose(
+        next_w.asnumpy()[start_cell_nudging:, :],
+        w_ref.asnumpy()[start_cell_nudging:, :],
+        rtol=1e-10,
+        atol=1e-12,
+    )
+    assert helpers.dallclose(rho_explicit_term.asnumpy(), z_rho_expl_ref.asnumpy())
+    assert helpers.dallclose(
+        exner_explicit_term.asnumpy(),
+        z_exner_expl_ref.asnumpy(),
+        rtol=1e-9,
+    )
+    assert helpers.dallclose(
+        next_rho.asnumpy()[start_cell_nudging:, :], rho_ref.asnumpy()[start_cell_nudging:, :]
+    )
+    assert helpers.dallclose(
+        next_exner.asnumpy()[start_cell_nudging:, :], exner_ref.asnumpy()[start_cell_nudging:, :]
+    )
+    assert helpers.dallclose(next_theta_v.asnumpy(), theta_v_ref.asnumpy())
+    # TODO: cannot find exit points for these two
```

Comment:
I think this has been fixed.

Reply:
are these in `savepoint_nonhydro_exit` ?